// @oceanprotocol/squid/dist/node/squid.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var Config_1 = __importDefault(require("./models/Config"));
exports.Config = Config_1.default;
var Account_1 = __importDefault(require("./ocean/Account"));
exports.Account = Account_1.default;
var DID_1 = __importDefault(require("./ocean/DID"));
exports.DID = DID_1.default;
var Ocean_1 = require("./ocean/Ocean");
exports.Ocean = Ocean_1.Ocean;
var Logger_1 = require("./utils/Logger");
exports.Logger = Logger_1.LoggerInstance;
var Keeper_1 = __importDefault(require("./keeper/Keeper"));
exports.Keeper = Keeper_1.default;
var templates = __importStar(require("./keeper/contracts/templates"));
exports.templates = templates;
var conditions = __importStar(require("./keeper/contracts/conditions"));
exports.conditions = conditions;
var utils = __importStar(require("./utils"));
exports.utils = utils;
__export(require("./ddo/DDO"));
var OceanAssets_1 = require("./ocean/OceanAssets");
exports.OrderProgressStep = OceanAssets_1.OrderProgressStep;
exports.CreateProgressStep = OceanAssets_1.CreateProgressStep;
var OceanVersions_1 = require("./ocean/OceanVersions");
exports.OceanPlatformTechStatus = OceanVersions_1.OceanPlatformTechStatus;
var templates_1 = require("./keeper/contracts/templates");
exports.AgreementTemplate = templates_1.AgreementTemplate;
var conditions_1 = require("./keeper/contracts/conditions");
exports.Condition = conditions_1.Condition;
exports.ConditionState = conditions_1.ConditionState;
